# Handling categorical data

In [3]:
import pandas as pd
adult_census = pd.read_csv("../datasets/adult-census.csv")

In [4]:
# remove a column
adult_census = adult_census.drop(columns=["education-num"])

In [5]:
# Split target from data
target_name = "class"
target = adult_census[target_name]
data = adult_census.drop(columns=[target_name])

In [6]:
# Get the datatypes for each column
data.dtypes

age                int64
workclass         object
education         object
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
dtype: object

### Select all the categorical columns using a sklearn function

In [7]:
from sklearn.compose import make_column_selector as selector

In [8]:
categorical_columns_selector = selector(dtype_include=object)
categorical_columns = categorical_columns_selector(data)
categorical_columns

['workclass',
 'education',
 'marital-status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native-country']

In [9]:
# subset all the categorical columns from the data
data_categorical = data[categorical_columns]
data_categorical.head()

workclass      education       marital-status          occupation  \
0     Private           11th        Never-married   Machine-op-inspct   
1     Private        HS-grad   Married-civ-spouse     Farming-fishing   
2   Local-gov     Assoc-acdm   Married-civ-spouse     Protective-serv   
3     Private   Some-college   Married-civ-spouse   Machine-op-inspct   
4           ?   Some-college        Never-married                   ?   

  relationship    race      sex  native-country  
0    Own-child   Black     Male   United-States  
1      Husband   White     Male   United-States  
2      Husband   White     Male   United-States  
3      Husband   Black     Male   United-States  
4    Own-child   White   Female   United-States

In [10]:
data_categorical["education"].value_counts()

education
HS-grad         15784
Some-college    10878
Bachelors        8025
Masters          2657
Assoc-voc        2061
11th             1812
Assoc-acdm       1601
10th             1389
7th-8th           955
Prof-school       834
9th               756
12th              657
Doctorate         594
5th-6th           509
1st-4th           247
Preschool          83
Name: count, dtype: int64

### Convert the categorical columns into numbers

This can be done with (here: two) encoders:
- Ordinal encoding -> for features that **have some order** in them, for instance T-shirt sizes (S, M, L, XL) or education (from low to high)
- one-hot encoding -> when features don't have any order. OneHot encoding introduces new columns (features) for every unique value in the category and assigns the sample with 0 or 1 to this feature. This is now binary, so neglects any order. The dimensions of the data (your table size) increase though.

### Ordinal encoding

In [11]:
from sklearn.preprocessing import OrdinalEncoder

education_column = data_categorical[["education"]]   # double [[ instead of [ make sure the output is a dataframe and not a series!
encoder = OrdinalEncoder().set_output(transform="pandas")
education_encoded = encoder.fit_transform(education_column)  # fit does not mean "fitting a model" here, but "fitting encoded categories"
education_encoded

education
0            1.0
1           11.0
2            7.0
3           15.0
4           15.0
...          ...
48837        7.0
48838       11.0
48839       11.0
48840       11.0
48841       11.0

[48842 rows x 1 columns]

In [12]:
# underscore after the attribute indicates that this is a private variable in the class 
# (although Python officially does not have private variables, the developers still sought for a way to indicate this and they chose for _ )
encoder.categories_

[array([' 10th', ' 11th', ' 12th', ' 1st-4th', ' 5th-6th', ' 7th-8th',
        ' 9th', ' Assoc-acdm', ' Assoc-voc', ' Bachelors', ' Doctorate',
        ' HS-grad', ' Masters', ' Preschool', ' Prof-school',
        ' Some-college'], dtype=object)]

In [13]:
data_categorical.head()

workclass      education       marital-status          occupation  \
0     Private           11th        Never-married   Machine-op-inspct   
1     Private        HS-grad   Married-civ-spouse     Farming-fishing   
2   Local-gov     Assoc-acdm   Married-civ-spouse     Protective-serv   
3     Private   Some-college   Married-civ-spouse   Machine-op-inspct   
4           ?   Some-college        Never-married                   ?   

  relationship    race      sex  native-country  
0    Own-child   Black     Male   United-States  
1      Husband   White     Male   United-States  
2      Husband   White     Male   United-States  
3      Husband   Black     Male   United-States  
4    Own-child   White   Female   United-States

In [14]:
data_encoded = encoder.fit_transform(data_categorical)
data_encoded.head()

workclass  education  marital-status  occupation  relationship  race  sex  \
0        4.0        1.0             4.0         7.0           3.0   2.0  1.0   
1        4.0       11.0             2.0         5.0           0.0   4.0  1.0   
2        2.0        7.0             2.0        11.0           0.0   4.0  1.0   
3        4.0       15.0             2.0         7.0           0.0   2.0  1.0   
4        0.0       15.0             4.0         0.0           3.0   4.0  0.0   

   native-country  
0            39.0  
1            39.0  
2            39.0  
3            39.0  
4            39.0

The whole data frame has now been converted to categorical data, as you can see in the 2 tables above (before and after the fit_transform)

# Exercise

In [15]:
data_categorical["marital-status"].value_counts()

marital-status
Married-civ-spouse       22379
Never-married            16117
Divorced                  6633
Separated                 1530
Widowed                   1518
Married-spouse-absent      628
Married-AF-spouse           37
Name: count, dtype: int64

In [16]:
encoder.categories_

[array([' ?', ' Federal-gov', ' Local-gov', ' Never-worked', ' Private',
        ' Self-emp-inc', ' Self-emp-not-inc', ' State-gov', ' Without-pay'],
       dtype=object),
 array([' 10th', ' 11th', ' 12th', ' 1st-4th', ' 5th-6th', ' 7th-8th',
        ' 9th', ' Assoc-acdm', ' Assoc-voc', ' Bachelors', ' Doctorate',
        ' HS-grad', ' Masters', ' Preschool', ' Prof-school',
        ' Some-college'], dtype=object),
 array([' Divorced', ' Married-AF-spouse', ' Married-civ-spouse',
        ' Married-spouse-absent', ' Never-married', ' Separated',
        ' Widowed'], dtype=object),
 array([' ?', ' Adm-clerical', ' Armed-Forces', ' Craft-repair',
        ' Exec-managerial', ' Farming-fishing', ' Handlers-cleaners',
        ' Machine-op-inspct', ' Other-service', ' Priv-house-serv',
        ' Prof-specialty', ' Protective-serv', ' Sales', ' Tech-support',
        ' Transport-moving'], dtype=object),
 array([' Husband', ' Not-in-family', ' Other-relative', ' Own-child',
        ' Unmarried'

In [ ]:
OrdinalEncoder?

## One-hot encoding

In [18]:
from sklearn.preprocessing import OneHotEncoder

In [19]:
education_column = data_categorical[["education"]]   # double [[ instead of [ make sure the output is a dataframe and not a series!
encoder = OneHotEncoder(sparse_output=False).set_output(transform="pandas")
education_encoded = encoder.fit_transform(education_column)  # fit does not mean "fitting a model" here, but "fitting encoded categories"
education_encoded

education_ 10th  education_ 11th  education_ 12th  education_ 1st-4th  \
0                  0.0              1.0              0.0                 0.0   
1                  0.0              0.0              0.0                 0.0   
2                  0.0              0.0              0.0                 0.0   
3                  0.0              0.0              0.0                 0.0   
4                  0.0              0.0              0.0                 0.0   
...                ...              ...              ...                 ...   
48837              0.0              0.0              0.0                 0.0   
48838              0.0              0.0              0.0                 0.0   
48839              0.0              0.0              0.0                 0.0   
48840              0.0              0.0              0.0                 0.0   
48841              0.0              0.0              0.0                 0.0   

       education_ 5th-6th  education_ 7th-8th  education_ 9th  \
0                     0.0                 0.0             0.0   
1                     0.0                 0.0             0.0   
2                     0.0                 0.0             0.0   
3                     0.0                 0.0             0.0   
4                     0.0                 0.0             0.0   
...                   ...                 ...             ...   
48837                 0.0                 0.0             0.0   
48838                 0.0                 0.0             0.0   
48839                 0.0                 0.0             0.0   
48840                 0.0                 0.0             0.0   
48841                 0.0                 0.0             0.0   

       education_ Assoc-acdm  education_ Assoc-voc  education_ Bachelors  \
0                        0.0                   0.0                   0.0   
1                        0.0                   0.0                   0.0   
2                        1.0                   0.0                   0.0   
3                        0.0                   0.0                   0.0   
4                        0.0                   0.0                   0.0   
...                      ...                   ...                   ...   
48837                    1.0                   0.0                   0.0   
48838                    0.0                   0.0                   0.0   
48839                    0.0                   0.0                   0.0   
48840                    0.0                   0.0                   0.0   
48841                    0.0                   0.0                   0.0   

       education_ Doctorate  education_ HS-grad  education_ Masters  \
0                       0.0                 0.0                 0.0   
1                       0.0                 1.0                 0.0   
2                       0.0                 0.0                 0.0   
3                       0.0                 0.0                 0.0   
4                       0.0                 0.0                 0.0   
...                     ...                 ...                 ...   
48837                   0.0                 0.0                 0.0   
48838                   0.0                 1.0                 0.0   
48839                   0.0                 1.0                 0.0   
48840                   0.0                 1.0                 0.0   
48841                   0.0                 1.0                 0.0   

       education_ Preschool  education_ Prof-school  education_ Some-college  
0                       0.0                     0.0                      0.0  
1                       0.0                     0.0                      0.0  
2                       0.0                     0.0                      0.0  
3                       0.0                     0.0                      1.0  
4                       0.0                     0.0                      1.0  
...                     .

### Explanation
here, a lot of columns have been added. Values are 1 and 0, where 1 means that the sample belongs to the category ("One = Hot")

In [20]:
data_encoded = encoder.fit_transform(data_categorical)
data_encoded.head()

workclass_ ?  workclass_ Federal-gov  workclass_ Local-gov  \
0           0.0                     0.0                   0.0   
1           0.0                     0.0                   0.0   
2           0.0                     0.0                   1.0   
3           0.0                     0.0                   0.0   
4           1.0                     0.0                   0.0   

   workclass_ Never-worked  workclass_ Private  workclass_ Self-emp-inc  \
0                      0.0                 1.0                      0.0   
1                      0.0                 1.0                      0.0   
2                      0.0                 0.0                      0.0   
3                      0.0                 1.0                      0.0   
4                      0.0                 0.0                      0.0   

   workclass_ Self-emp-not-inc  workclass_ State-gov  workclass_ Without-pay  \
0                          0.0                   0.0                     0.0   
1                          0.0                   0.0                     0.0   
2                          0.0                   0.0                     0.0   
3                          0.0                   0.0                     0.0   
4                          0.0                   0.0                     0.0   

   education_ 10th  ...  native-country_ Portugal  \
0              0.0  ...                       0.0   
1              0.0  ...                       0.0   
2              0.0  ...                       0.0   
3              0.0  ...                       0.0   
4              0.0  ...                       0.0   

   native-country_ Puerto-Rico  native-country_ Scotland  \
0                          0.0                       0.0   
1                          0.0                       0.0   
2                          0.0                       0.0   
3                          0.0                       0.0   
4                          0.0                       0.0   

   native-country_ South  native-country_ Taiwan  native-country_ Thailand  \
0                    0.0                     0.0                       0.0   
1                    0.0                     0.0                       0.0   
2                    0.0                     0.0                       0.0   
3                    0.0                     0.0                       0.0   
4                    0.0                     0.0                       0.0   

   native-country_ Trinadad&Tobago  native-country_ United-States  \
0                              0.0                            1.0   
1                              0.0                            1.0   
2                              0.0                            1.0   
3                              0.0                            1.0   
4                              0.0                            1.0   

   native-country_ Vietnam  native-country_ Yugoslavia  
0                      0.0                         0.0  
1                      0.0                         0.0  
2                      0.0                         0.0  
3                      0.0                         0.0  
4                      0.0                         0.0  

[5 rows x 102 columns]

#### Now we have 102 instead of 8 columns! 

When to use one-hot encoding or ordinal encoding? 
Depends on the machine learning you want to apply. Some models just want a label assigned to the category and neglect the order. Then, ordinary encoding is fine.
Other models assume that there is some order in the data, which forces you to **not** use ordinal encoding.

## Exercise One-Hot encoding

### Exercise: The impact of using integer encoding for with logistic regression (groups of 2, 15min)


Goal: understand the impact of arbitrary integer encoding for categorical variables with linear classification such as logistic regression.

We keep using the `adult_census` data set already loaded in the code before. Recall that `target` contains the variable we want to predict and `data` contains the features.

If you need to re-load the data, you can do it as follows:

```python
import pandas as pd

adult_census = pd.read_csv("../datasets/adult-census.csv")
target_name = "class"
target = adult_census[target_name]
data = adult_census.drop(columns=[target_name, "education-num"])
```



**Q1 Build a scikit-learn pipeline composed of an `OrdinalEncoder` and a `LogisticRegression` classifier**

You'll need the following, already loaded modules:

```python
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import OrdinalEncoder
from sklearn.linear_model import LogisticRegression
```

Because OrdinalEncoder can raise errors if it sees an unknown category at prediction time, you can set the handle_unknown="use_encoded_value" and unknown_value parameters. You can refer to the [scikit-learn documentation](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OrdinalEncoder.html) for more details regarding these parameters.


**Q2 Evaluate the model with cross-validation.**

You'll need the following, already loaded modules:

```python
from sklearn.model_selection import cross_validate

```

**Q3 Repeat the previous steps using an `OneHotEncoder` instead of an `OrdinalEncoder`**

You'll need the following, already loaded modules:

```python
from sklearn.preprocessing import OneHotEncoder

```

**Q0 Select columns containing strings**
Use `sklearn.compose.make_column_selector` to automatically select columns containing strings that correspond to categorical features in our dataset.


In [32]:
# Split target from data
target_name = "class"
target = adult_census[target_name]
data = adult_census.drop(columns=[target_name])

categorical_columns_selector = selector(dtype_include=object)
categorical_columns = categorical_columns_selector(data)
data_cat = data[categorical_columns]

In [33]:
data_cat.head()

workclass      education       marital-status          occupation  \
0     Private           11th        Never-married   Machine-op-inspct   
1     Private        HS-grad   Married-civ-spouse     Farming-fishing   
2   Local-gov     Assoc-acdm   Married-civ-spouse     Protective-serv   
3     Private   Some-college   Married-civ-spouse   Machine-op-inspct   
4           ?   Some-college        Never-married                   ?   

  relationship    race      sex  native-country  
0    Own-child   Black     Male   United-States  
1      Husband   White     Male   United-States  
2      Husband   White     Male   United-States  
3      Husband   Black     Male   United-States  
4    Own-child   White   Female   United-States

**Q1 Build a scikit-learn pipeline composed of an `OrdinalEncoder` and a `LogisticRegression` classifier**

In [54]:
import numpy as np
from sklearn.preprocessing import OrdinalEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import make_pipeline

# input = all the preprocessing steps + the model to fit
model = make_pipeline(OrdinalEncoder(handle_unknown="use_encoded_value", unknown_value=-1), LogisticRegression())

**Q2 Evaluate the model with cross-validation.**

In [55]:
from sklearn.model_selection import cross_validate

In [56]:
cv_result = cross_validate(model, data_categorical, target, cv=5)
# this will cause a lot of warnings, but the calculation continues and provides results

/home/maarten/miniconda3/envs/scikit-learn-course/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/home/maarten/miniconda3/envs/scikit-learn-course/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-

In [57]:
cv_result

{'fit_time': array([0.48274994, 0.27955008, 0.2761867 , 0.32835507, 0.30605483]),
 'score_time': array([0.04466343, 0.01855779, 0.01972818, 0.01931572, 0.0176096 ]),
 'test_score': array([0.75514382, 0.75555328, 0.75573301, 0.75317363, 0.75788288])}

**Q3 Repeat the previous steps using an `OneHotEncoder` instead of an `OrdinalEncoder`**

In [53]:
from sklearn.preprocessing import OneHotEncoder

# Hint from instructor Johan: use "handle_unknown="ignore"
# This is because "Holland" occurs only once in native country column
# So it can happen that in one of the iterations of the crossvalidation the value Holland is missing from the test set
# And the train and test set contain different values.
model2 = make_pipeline(OneHotEncoder(sparse_output=False, handle_unknown="ignore").set_output(transform="pandas"), LogisticRegression())

In [50]:
cv_result2 = cross_validate(model2, data_categorical, target, cv=5)

In [51]:
cv_result2

{'fit_time': array([0.87303281, 0.74400687, 1.55929613, 1.18916011, 0.91958213]),
 'score_time': array([0.14119554, 0.07688665, 0.0517993 , 0.10233331, 0.09305358]),
 'test_score': array([0.83232675, 0.83570478, 0.82831695, 0.83292383, 0.83497133])}